File: languages/python/02-core/file-io-basics/example/main.py
# This example shows reading plain-text files, parsing rows, and writing clear results.
# In Python, the example favors direct readable steps while keeping validation visible.

from pathlib import Path
import tempfile


# Define the reusable pieces first so the main flow can focus on one observable scenario.
def parse_score_row(line):
    # Build the sample state first, then let the later output confirm the behavior step by step.
    parts = line.strip().split()
    if len(parts) != 2:
        return None

    try:
        score = int(parts[1])
    except ValueError:
        return None

    return parts[0], score


# Run one deterministic scenario so the console output makes reading plain-text files, parsing rows,
# and writing clear results easy to verify.
def main():
    base_dir = Path(tempfile.gettempdir()) / "learn-lang-file-io-python"
    base_dir.mkdir(parents=True, exist_ok=True)
    input_path = base_dir / "scores.txt"
    output_path = base_dir / "summary.txt"

    if not input_path.exists():
        input_path.write_text("ana 90\nbob 82\ninvalid row\ncarla 95\n", encoding="utf-8")

    valid_rows = 0
    score_sum = 0

    with input_path.open("r", encoding="utf-8") as source:
        for raw_line in source:
            parsed = parse_score_row(raw_line)
            if parsed is None:
                continue

            name, score = parsed
            valid_rows += 1
            score_sum += score
            print(f"{name} -> {score}")

    if valid_rows == 0:
        print("No valid rows found.")
        return

    average = score_sum / valid_rows
    output_path.write_text(f"Rows: {valid_rows}\nAverage: {average:.2f}\n", encoding="utf-8")
    print(f"Summary written to {output_path}")


if __name__ == "__main__":
    main()

File: languages/go/02-core/file-io-basics/example/main.go
// Module focus: Reading plain-text files, parsing rows, and writing clear results.
// Why it matters: practicing file io basics patterns makes exercises and checkpoints easier to reason about.

package main

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func main() {
	// Explicit arguments make missing input observable instead of creating it silently.
	inputPath := filepath.Join("example", "fixtures", "scores.txt")
	outputPath := filepath.Join("build", "report.txt")
	if len(os.Args) > 1 {
		inputPath = os.Args[1]
	}
	if len(os.Args) > 2 {
		outputPath = os.Args[2]
	}

	file, err := os.Open(inputPath)
	if err != nil {
		fmt.Printf("Could not open %s: %v\n", inputPath, err)
		return
	}
	defer file.Close()

	// Keep accepted and rejected records separate so the report can explain both.
	var records []scoreRecord
	invalidRows := 0
	// Blank lines are formatting noise; malformed non-blank rows count as rejected data.
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		if strings.TrimSpace(scanner.Text()) == "" {
			continue
		}
		record, ok := parseScoreRow(scanner.Text())
		if ok {
			records = append(records, record)
		} else {
			invalidRows++
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Printf("Could not read input file: %v\n", err)
		return
	}

	// Build and persist one deterministic learner-visible report.
	report := buildReport(records, invalidRows)
	if err := os.MkdirAll(filepath.Dir(outputPath), 0o755); err != nil {
		fmt.Printf("Could not create output directory: %v\n", err)
		return
	}
	if err := os.WriteFile(outputPath, []byte(report+"\n"), 0o644); err != nil {
		fmt.Printf("Could not create %s: %v\n", outputPath, err)
		return
	}

	// Print the same report so file and console results are easy to compare.
	// The explicit path also tells learners exactly where the generated artifact lives.
	fmt.Printf("Report file: %s\n", outputPath)
	fmt.Println(report)
}
